path-tracker-pipelines/src/main/java/org/example/merger/KafkaConsumerThread.java

```diff
@@ -16,83 +16,84 @@
  * limitations under the License.
  */
 
-package org.example;
+package org.example.merger;
 
 import org.apache.kafka.clients.consumer.Consumer;
 import org.apache.kafka.clients.consumer.ConsumerConfig;
 import org.apache.kafka.clients.consumer.ConsumerRecord;
 import org.apache.kafka.clients.consumer.ConsumerRecords;
 import org.apache.kafka.clients.consumer.KafkaConsumer;
 import org.apache.kafka.common.serialization.StringDeserializer;
+import org.example.datasource.DecorateRecord;
+import org.example.utils.RecordSerdes;
 
+import java.nio.charset.StandardCharsets;
 import java.time.Duration;
 import java.util.Collections;
 import java.util.Properties;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentLinkedQueue;
-import java.util.concurrent.atomic.AtomicInteger;
-import java.util.concurrent.atomic.AtomicLong;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
 
 public class KafkaConsumerThread implements Runnable{
         private volatile boolean running = true;
         private int partitionCount;
         Consumer<String, String> consumer;
-        private ConcurrentLinkedQueue<kafkaMessage> partitionQueue [];
-        String pattern = "(\\d+)-(\\d+)"; // detects things in the pattern %d-%d
-        Pattern regex;
-        Matcher matcher;
-        AtomicLong watermark;
+        private final ConcurrentLinkedQueue<kafkaMessage> [] partitionQueue;
+        ConcurrentHashMap<Integer, Long> watermarks;
+        long recordsReceived = 0;
 
 
-    public KafkaConsumerThread(String bootstrapServer , int partitionCount, ConcurrentLinkedQueue<kafkaMessage>[] queue, AtomicLong watermark) {
+    public KafkaConsumerThread(String bootstrapServer , int partitionCount, ConcurrentLinkedQueue<kafkaMessage>[] queue,  ConcurrentHashMap<Integer, Long> watermarks) {
             this.partitionCount = partitionCount;
             Properties consumerProps = new Properties();
-            consumerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServer); // Replace with your Kafka broker addresses
-            consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, "test_group_id"); // Consumer group ID
+            consumerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServer);
+            consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, "test_group_id");
             consumerProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
             consumerProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
             this.consumer = new KafkaConsumer<>(consumerProps);
-            this.consumer.subscribe(Collections.singletonList("test_topic")); // Need to update with watermark topic
+            this.consumer.subscribe(Collections.singletonList("test_topic"));
```

Comment:
Might want to avoid hardcoding this value to avoid data interference between runs. I'm thinking about using a random UUID for test_topic for each run. The value should be read from argument list.